src/main/java/org/jsoup/select/StructuralEvaluator.java

```diff
@@ -24,18 +25,9 @@ public StructuralEvaluator(Evaluator evaluator) {
 
     boolean memoMatches(final Element root, final Element element) {
         // not using computeIfAbsent, as the lambda impl requires a new Supplier closure object on every hit: tons of GC
-        IdentityHashMap<Element, IdentityHashMap<Element, Boolean>> rootMemo = threadMemo.get();
-        IdentityHashMap<Element, Boolean> memo = rootMemo.get(root);
-        if (memo == null) {
-            memo = new IdentityHashMap<>();
-            rootMemo.put(root, memo);
-        }
-        Boolean matches = memo.get(element);
-        if (matches == null) {
-            matches = evaluator.matches(root, element);
-            memo.put(element, matches);
-        }
-        return matches;
+        Map<Element, IdentityHashMap<Element, Boolean>> rootMemo = threadMemo.get();
```

Comment:
Please see the code comment in line 27 -- I tried this before and under profiling, there is a ton of garbage generation for the lambda closure; a new object on every method call.

Perhaps make the supplier a static class vs a lambda?